client/session.go

```diff
@@ -563,3 +595,945 @@ func (s *session) Close() error {
+		// ordered by time so must be first of each
+		currStart = currStart[:0]
+		for _, blocksMetadata := range perPeerBlocksMetadata {
+			if len(blocksMetadata.unselectedBlocks()) == 0 {
+				// No unselected blocks
+				continue
+			}
+
+			unselected := blocksMetadata.unselectedBlocks()
+			if !unselected[0].start.Equal(earliestStart) {
+				// Not the same block
+				continue
+			}
+
+			currStart = append(currStart, blocksMetadata)
+		}
+
+		if len(currStart) == 0 {
+			// No more blocks to select from any peers
+			break
+		}
+
+		// Only select from peers not already attempted
+		currEligible = currStart[:]
+		currID := currStart[0].id
+		for i := len(currEligible) - 1; i >= 0; i-- {
+			unselected := currEligible[i].unselectedBlocks()
+			if unselected[0].reattempt.attempt == 0 {
+				// Not attempted yet
+				continue
+			}
+
+			// Check if eligible
+			if unselected[0].reattempt.hasAttemptedPeer(currEligible[i].peer) {
+				// Swap current entry to tail
+				blocksMetadatas(currEligible).swap(i, len(currEligible)-1)
+				// Trim newly last entry
+				currEligible = currEligible[:len(currEligible)-1]
+				continue
+			}
+		}
+
+		if len(currEligible) == 0 {
+			// No current eligible peers to select from
+			s.log.WithFields(
+				xlog.NewLogField("id", currID),
+				xlog.NewLogField("start", earliestStart),
+				xlog.NewLogField("attempted", currStart[0].unselectedBlocks()[0].reattempt.attempt),
+			).Error("retries failed for streaming blocks from peers")
+
+			// Remove the block from all peers
+			for i := range currStart {
+				blocksLen := len(currStart[i].blocks)
+				idx := currStart[i].idx
+				tailIdx := blocksLen - 1
+				currStart[i].blocks[idx], currStart[i].blocks[tailIdx] = currStart[i].blocks[tailIdx], currStart[i].blocks[idx]
+				currStart[i].blocks = currStart[i].blocks[:blocksLen-1]
+			}
+			continue
+		}
+
+		// Determine maximum size
+		currMaxSize := int64(math.MinInt64)
+		for i := 0; i < len(currEligible); i++ {
+			unselected := currEligible[i].unselectedBlocks()
+			// Track maximum size
+			if unselected[0].size > currMaxSize {
+				currMaxSize = unselected[0].size
+			}
+		}
+
+		// Only select from those with the maximum size
+		for i := len(currEligible) - 1; i >= 0; i-- {
+			unselected := currEligible[i].unselectedBlocks()
+			if unselected[0].size < currMaxSize {
+				// Swap current entry to tail
+				blocksMetadatas(currEligible).swap(i, len(currEligible)-1)
+				// Trim newly last entry
+				currEligible = currEligible[:len(currEligible)-1]
+			}
+		}
+
+		// Order by least outstanding blocks being fetched
+		blocksMetadataQueues = blocksMetadataQueues[:0]
+		for i := range currEligible {
+			insert := blocksMetadataQueue{
+				blocksMetadata: currEligible[i],
+				queue:          peerQueues.findQueue(currEligible[i].peer),
+			}
+			blocksMetadataQueues = append(blocksMetadataQueues, insert)
+		}
+		sort.Stable(blocksMetadatasQueuesByOutstanding(blocksMetadataQueues))
```

Comment:
For predictable test assertions, also the first ordered queue if they are all equal will get it every time which makes for more predictable outcomes when observing behavior.